File: search/binary_search/javascript/binary_search.js
(function(){
"use strict"

const binarySearch = function(arr, num, start = 0, end = (arr.length - 1)) {
    let middle = Math.floor((start + end) / 2);
    switch (true) {
        case arr[middle] === num:
            return arr[middle];
        case end - start === 0:
            return false;
        case arr[middle] < num:
            return binarySearch(arr, num, middle + 1, end);
        case arr[middle] > num:
            return binarySearch(arr, num, start, middle);
    }
    return middle;
}

module.exports = binarySearch;

})();

File: search/binary_search/swift/BinarySearch.swift
func binarySearch(find: Int, arr: [Int], start: Int = 0, end: Int?) -> Int? {
    let endValue = end ?? arr.count - 1
    let middle = (endValue - start)/2 + start
    if arr[middle] == find {
        return middle
    } else {
        if endValue == start {
            return nil
        } else {
            if find > arr[middle] {
                return binarySearch(find: find, arr: arr, start: middle + 1, end: endValue)
            } else {
                return binarySearch(find: find, arr: arr, start: start, end: middle - 1)
            }
        }
    }
}
